packages/jsii-pacmak/lib/targets/dotnet/runtime-type-checking.ts

```diff
@@ -0,0 +1,242 @@
+import { CodeMaker } from 'codemaker';
+import { createHash } from 'crypto';
+import { Parameter, TypeReference } from 'jsii-reflect';
+
+import { DotNetTypeResolver } from './dotnettyperesolver';
+import { DotNetNameUtils } from './nameutils';
+
+export class ParameterValidator {
+  public static forParameters(
+    parameters: readonly Parameter[],
+    nameUtils: DotNetNameUtils,
+    { noMangle }: { readonly noMangle: boolean },
+  ): ParameterValidator | undefined {
+    if (parameters.length === 0) {
+      return undefined;
+    }
+    const parameterValidations = new Map<Parameter, Validation[]>();
+
+    for (const param of parameters) {
+      const expr = noMangle
+        ? param.name
+        : nameUtils.convertParameterName(param.name);
+      const argName = `nameof(${expr})`;
+
+      const validations = new Array<Validation>();
+      const validation = Validation.forTypeReference(
+        argName,
+        expr,
+        `${noMangle ? '' : 'argument '}{${argName}}`,
+        param.type,
+        param.optional,
+      );
+      if (validation) {
+        validations.push(validation);
+      }
+
+      if (validations.length !== 0) {
+        parameterValidations.set(param, validations);
+      }
+    }
+
+    if (parameterValidations.size === 0) {
+      return undefined;
+    }
+    return new ParameterValidator(parameterValidations);
+  }
+
+  private constructor(
+    private readonly validations: ReadonlyMap<Parameter, readonly Validation[]>,
+  ) {}
+
+  public emit(code: CodeMaker, resolver: DotNetTypeResolver): void {
+    for (const [_parameter, validations] of this.validations) {
+      for (const validation of validations) {
+        validation.emit(code, resolver);
+      }
+    }
+  }
+}
+
+abstract class Validation {
```

Comment:
Should I use this approach in the Java code gen as well, or stick to the previous way?

Reply:
You can if you want, but you can also migrate in a separate PR.